File: sociallang/lang/parser.py
from __future__ import annotations

from .ast_nodes import (
    Attr, AssignStmt, BinaryOp, Block, BoolLit, BreakStmt, Call, DictLit, Expr, ExprStmt,
    FnDecl, ForStmt, Identifier, IfStmt, Index, LetStmt, ListLit, MemoryDecl, NullLit,
    NumberLit, PhaseDecl, ReturnStmt, RoleDecl, RunStmt, SimDecl, Stmt, StringLit, UnaryOp,
    WhileStmt,
)
from .lexer import Token, tokenize


class ParseError(Exception):
    pass


class Parser:
    def __init__(self, tokens: list[Token]):
        self.tokens = tokens
        self.pos = 0

    # -- token stream helpers --

    def _peek(self) -> Token:
        return self.tokens[self.pos]

    def _advance(self) -> Token:
        tok = self.tokens[self.pos]
        if tok.kind != "EOF":
            self.pos += 1
        return tok

    def _check(self, kind: str, value: str | None = None) -> bool:
        tok = self._peek()
        return tok.kind == kind and (value is None or tok.value == value)

    def _match(self, kind: str, value: str | None = None) -> Token | None:
        if self._check(kind, value):
            return self._advance()
        return None

    def _expect(self, kind: str, value: str | None = None) -> Token:
        tok = self._match(kind, value)
        if tok is None:
            cur = self._peek()
            expected = value or kind
            raise ParseError(f"line {cur.line}: expected {expected!r}, got {cur.value!r} ({cur.kind})")
        return tok

    def _is_kw(self, word: str) -> bool:
        return self._check("KEYWORD", word)

    # -- entry point --

    def parse_program(self) -> SimDecl:
        self._expect("KEYWORD", "sim")
        name = self._expect("IDENT").value
        self._expect("SYMBOL", "{")

        sim = SimDecl(name=name, agents_min=0, agents_max=0)
        while not self._check("SYMBOL", "}"):
            if self._is_kw("agents"):
                self._advance()
                self._expect("SYMBOL", ":")
                lo = int(float(self._expect("NUMBER").value))
                hi = lo
                if self._match("SYMBOL", ".."):
                    hi = int(float(self._expect("NUMBER").value))
                sim.agents_min, sim.agents_max = lo, hi
            elif self._is_kw("memory"):
                sim.memory_decls.append(self._parse_memory_decl())
            elif self._is_kw("role"):
                sim.roles.append(self._parse_role_decl())
            elif self._is_kw("fn"):
                sim.fns.append(self._parse_fn_decl())
            elif self._is_kw("phase"):
                sim.phases.append(self._parse_phase_decl())
            elif self._is_kw("win_condition"):
                self._advance()
                sim.win_condition = self._parse_block()
            elif self._is_kw("loop"):
                self._advance()
                sim.loop = self._parse_block()
            else:
                cur = self._peek()
                raise ParseError(f"line {cur.line}: unexpected token {cur.value!r} in sim body")
        self._expect("SYMBOL", "}")
        return sim

    def _parse_memory_decl(self) -> MemoryDecl:
        self._advance()  # 'memory'
        name = self._expect("IDENT").value
        params: list[str] = []
        if self._match("SYMBOL", "("):
            while not self._check("SYMBOL", ")"):
                params.append(self._expect("IDENT").value)
                if not self._match("SYMBOL", ","):
                    break
            self._expect("SYMBOL", ")")
        body = self._parse_block()
        return MemoryDecl(name=name, params=params, body=body)

    def _parse_role_decl(self) -> RoleDecl:
        self._advance()  # 'role'
        name = self._expect("IDENT").value
        self._expect("SYMBOL", "{")

        team = name
        memory_name: str | None = None
        memory_args: list[Expr] = []
        sees = "none"
        count: int | None = None

        while not self._check("SYMBOL", "}"):
            field_name = self._advance().value  # IDENT or KEYWORD (e.g. "memory")
            self._expect("SYMBOL", ":")
            if field_name == "team":
                team = self._expect("STRING").value
            elif field_name == "memory":
                memory_name = self._expect("IDENT").value
                if self._match("SYMBOL", "("):
                    while not self._check("SYMBOL", ")"):
                        memory_args.append(self._parse_expr())
                        if not self._match("SYMBOL", ","):
                            break
                    self._expect("SYMBOL", ")")
            elif field_name == "sees":
                sees = self._advance().value
            elif field_name == "count":
                tok = self._peek()
                if tok.kind == "IDENT" and tok.value == "remainder":
                    self._advance()
                    count = None
                else:
                    count = int(float(self._expect("NUMBER").value))
            else:
                raise ParseError(f"unknown role field {field_name!r}")
            self._match("SYMBOL", ",")

        self._expect("SYMBOL", "}")
        return RoleDecl(name=name, team=team, memory_name=memory_name, memory_args=memory_args, sees=sees, count=count)

    def _parse_fn_decl(self) -> FnDecl:
        self._advance()  # 'fn'
        name = self._expect("IDENT").value
        self._expect("SYMBOL", "(")
        params: list[str] = []
        while not self._check("SYMBOL", ")"):
            params.append(self._expect("IDENT").value)
            if not self._match("SYMBOL", ","):
                break
        self._expect("SYMBOL", ")")
        body = self._parse_block()
        return FnDecl(name=name, params=params, body=body)

    def _parse_phase_decl(self) -> PhaseDecl:
        self._advance()  # 'phase'
        name = self._expect("IDENT").value
        body = self._parse_block()
        return PhaseDecl(name=name, body=body)

    # -- statements --

    def _parse_block(self) -> Block:
        self._expect("SYMBOL", "{")
        statements: list[Stmt] = []
        while not self._check("SYMBOL", "}"):
            statements.append(self._parse_statement())
        self._expect("SYMBOL", "}")
        return Block(statements)

    def _parse_statement(self) -> Stmt:
        if self._is_kw("let"):
            self._advance()
            name = self._expect("IDENT").value
            self._expect("SYMBOL", "=")
            value = self._parse_expr()
            return LetStmt(name, value)
        if self._is_kw("if"):
            return self._parse_if()
        if self._is_kw("while"):
            self._advance()
            cond = self._parse_expr()
            body = self._parse_block()
            return WhileStmt(cond, body)
        if self._is_kw("for"):
            self._advance()
            var = self._expect("IDENT").value
            self._expect("KEYWORD", "in")
            iterable = self._parse_expr()
            body = self._parse_block()
            return ForStmt(var, iterable, body)
        if self._is_kw("return"):
            self._advance()
            if self._check("SYMBOL", "}"):
                return ReturnStmt(None)
            return ReturnStmt(self._parse_expr())
        if self._is_kw("break"):
            self._advance()
            return BreakStmt()
        if self._is_kw("run"):
            self._advance()
            name = self._expect("IDENT").value
            return RunStmt(name)

        expr = self._parse_expr()
        if self._match("SYMBOL", "="):
            value = self._parse_expr()
            return AssignStmt(expr, value)
        return ExprStmt(expr)

    def _parse_if(self) -> IfStmt:
        self._advance()  # 'if'
        cond = self._parse_expr()
        then_block = self._parse_block()
        else_block = None
        if self._match("KEYWORD", "else"):
            if self._is_kw("if"):
                else_block = self._parse_if()
            else:
                else_block = self._parse_block()
        return IfStmt(cond, then_block, else_block)

    # -- expressions (precedence climbing) --

    def _parse_expr(self) -> Expr:
        return self._parse_or()

    def _parse_or(self) -> Expr:
        left = self._parse_and()
        while self._match("KEYWORD", "or"):
            left = BinaryOp("or", left, self._parse_and())
        return left

    def _parse_and(self) -> Expr:
        left = self._parse_not()
        while self._match("KEYWORD", "and"):
            left = BinaryOp("and", left, self._parse_not())
        return left

    def _parse_not(self) -> Expr:
        if self._match("KEYWORD", "not"):
            return UnaryOp("not", self._parse_not())
        return self._parse_comparison()

    _CMP_OPS = {"==", "!=", "<", ">", "<=", ">="}

    def _parse_comparison(self) -> Expr:
        left = self._parse_additive()
        while self._peek().kind == "SYMBOL" and self._peek().value in self._CMP_OPS:
            op = self._advance().value
            left = BinaryOp(op, left, self._parse_additive())
        return left

    def _parse_additive(self) -> Expr:
        left = self._parse_multiplicative()
        while self._peek().kind == "SYMBOL" and self._peek().value in ("+", "-"):
            op = self._advance().value
            left = BinaryOp(op, left, self._parse_multiplicative())
        return left

    def _parse_multiplicative(self) -> Expr:
        left = self._parse_unary()
        while self._peek().kind == "SYMBOL" and self._peek().value in ("*", "/", "%"):
            op = self._advance().value
            left = BinaryOp(op, left, self._parse_unary())
        return left

    def _parse_unary(self) -> Expr:
        if self._match("SYMBOL", "-"):
            return UnaryOp("-", self._parse_unary())
        return self._parse_postfix()

    def _parse_postfix(self) -> Expr:
        expr = self._parse_primary()
        while True:
            if self._match("SYMBOL", "."):
                name = self._expect("IDENT").value
                expr = Attr(expr, name)
            elif self._match("SYMBOL", "("):
                args: list[Expr] = []
                kwargs: dict[str, Expr] = {}
                while not self._check("SYMBOL", ")"):
                    if self._peek().kind == "IDENT" and self.tokens[self.pos + 1].kind == "SYMBOL" \
                            and self.tokens[self.pos + 1].value == "=":
                        key = self._advance().value
                        self._advance()  # '='
                        kwargs[key] = self._parse_expr()
                    else:
                        args.append(self._parse_expr())
                    if not self._match("SYMBOL", ","):
                        break
                self._expect("SYMBOL", ")")
                expr = Call(expr, args, kwargs)
            elif self._match("SYMBOL", "["):
                key = self._parse_expr()
                self._expect("SYMBOL", "]")
                expr = Index(expr, key)
            else:
                break
        return expr

    def _parse_primary(self) -> Expr:
        tok = self._peek()
        if tok.kind == "NUMBER":
            self._advance()
            return NumberLit(float(tok.value))
        if tok.kind == "STRING":
            self._advance()
            return StringLit(tok.value)
        if tok.kind == "KEYWORD" and tok.value == "true":
            self._advance()
            return BoolLit(True)
        if tok.kind == "KEYWORD" and tok.value == "false":
            self._advance()
            return BoolLit(False)
        if tok.kind == "KEYWORD" and tok.value == "null":
            self._advance()
            return NullLit()
        if tok.kind == "IDENT":
            self._advance()
            return Identifier(tok.value)
        if self._match("SYMBOL", "("):
            inner = self._parse_expr()
            self._expect("SYMBOL", ")")
            return inner
        if self._match("SYMBOL", "["):
            items: list[Expr] = []
            while not self._check("SYMBOL", "]"):
                items.append(self._parse_expr())
                if not self._match("SYMBOL", ","):
                    break
            self._expect("SYMBOL", "]")
            return ListLit(items)
        if self._match("SYMBOL", "{"):
            pairs: list[tuple[Expr, Expr]] = []
            while not self._check("SYMBOL", "}"):
                key = self._parse_expr()
                self._expect("SYMBOL", ":")
                value = self._parse_expr()
                pairs.append((key, value))
                if not self._match("SYMBOL", ","):
                    break
            self._expect("SYMBOL", "}")
            return DictLit(pairs)

        raise ParseError(f"line {tok.line}: unexpected token {tok.value!r} ({tok.kind})")


def parse(source: str) -> SimDecl:
    return Parser(tokenize(source)).parse_program()

File: cpp/src/engine/parser.cpp
// Recursive-descent parser, a line-for-line port of sociallang/lang/parser.py
// (same precedence ladder: or < and < not < comparison < additive <
// multiplicative < unary < postfix < primary), plus a link pass that resolves
// every call and `run` to an index.
#include "engine/parser.h"

#include <cmath>
#include <unordered_map>

#include "engine/builtins.h"
#include "engine/lexer.h"

namespace sl {

SymbolTable::SymbolTable() {
  static const char* kFixed[S_COUNT_] = {
      "seat", "role", "team", "alive", "model", "death_cause", "x", "y", "location",
      "persona", "plan", "subplan", "text", "author", "kind", "seq", "round",
      "importance", "id", "type", "tag", "capacity",
      "temperature", "max_tokens", "threshold", "goal", "steps", "chunks", "topic",
      "max_turns", "cause", "radius",
      "agents", "events", "query",
  };
  for (const char* name : kFixed) intern(name);
}

int SymbolTable::intern(const std::string& name) {
  auto it = ids_.find(name);
  if (it != ids_.end()) return it->second;
  int id = static_cast<int>(names_.size());
  names_.push_back(name);
  ids_.emplace(name, id);
  return id;
}

namespace {

using ExprP = std::unique_ptr<Expr>;
using StmtP = std::unique_ptr<Stmt>;

class Parser {
 public:
  Parser(std::vector<Token> tokens, Program& prog) : toks_(std::move(tokens)), prog_(prog) {}

  void parseProgram() {
    expect(TK::Keyword, "sim");
    prog_.name = expect(TK::Ident).value;
    expect(TK::Symbol, "{");
    while (!check(TK::Symbol, "}")) {
      if (isKw("agents")) {
        advance();
        expect(TK::Symbol, ":");
        int lo = intLit();
        int hi = lo;
        if (match(TK::Symbol, "..")) hi = intLit();
        prog_.agentsMin = lo;
        prog_.agentsMax = hi;
      } else if (isKw("memory")) {
        prog_.memoryDecls.push_back(parseFnLike(/*requireParens=*/false));
      } else if (isKw("role")) {
        prog_.roles.push_back(parseRole());
      } else if (isKw("fn")) {
        prog_.fns.push_back(parseFnLike(/*requireParens=*/true));
      } else if (isKw("phase")) {
        advance();
        PhaseDecl p;
        p.name = expect(TK::Ident).value;
        p.body = parseBlock();
        prog_.phases.push_back(std::move(p));
      } else if (isKw("win_condition")) {
        advance();
        prog_.winCondition = parseBlock();
      } else if (isKw("loop")) {
        advance();
        prog_.loop = parseBlock();
      } else if (isKw("world")) {
        prog_.world = parseWorld();
      } else {
        fail("unexpected token '" + peek().value + "' in sim body");
      }
    }
    expect(TK::Symbol, "}");
  }

 private:
  std::vector<Token> toks_;
  size_t pos_ = 0;
  Program& prog_;

  const Token& peek(size_t ahead = 0) const {
    size_t i = pos_ + ahead;
    return toks_[i < toks_.size() ? i : toks_.size() - 1];
  }
  const Token& advance() {
    const Token& t = toks_[pos_];
    if (t.kind != TK::End) ++pos_;
    return t;
  }
  bool check(TK kind, const char* value = nullptr) const {
    const Token& t = peek();
    return t.kind == kind && (!value || t.value == value);
  }
  bool match(TK kind, const char* value = nullptr) {
    if (!check(kind, value)) return false;
    advance();
    return true;
  }
  const Token& expect(TK kind, const char* value = nullptr) {
    if (!check(kind, value)) {
      static const char* kNames[] = {"NUMBER", "STRING", "KEYWORD", "IDENT", "SYMBOL", "EOF"};
      std::string want = value ? value : kNames[static_cast<int>(kind)];
      fail("expected '" + want + "', got '" + peek().value + "'");
    }
    return advance();
  }
  bool isKw(const char* w) const { return check(TK::Keyword, w); }
  [[noreturn]] void fail(const std::string& msg) const {
    throw ParseError("line " + std::to_string(peek().line) + ": " + msg);
  }
  int intLit() { return static_cast<int>(std::trunc(std::stod(expect(TK::Number).value))); }

  ExprP node(EK k, int line) {
    auto e = std::make_unique<Expr>();
    e->k = k;
    e->line = line;
    return e;
  }

  FnDecl parseFnLike(bool requireParens) {
    advance();
    FnDecl f;
    f.name = expect(TK::Ident).value;
    bool open = requireParens ? (expect(TK::Symbol, "("), true) : match(TK::Symbol, "(");
    if (open) {
      while (!check(TK::Symbol, ")")) {
        f.params.push_back(prog_.syms.intern(expect(TK::Ident).value));
        if (!match(TK::Symbol, ",")) break;
      }
      expect(TK::Symbol, ")");
    }
    f.body = parseBlock();
    return f;
  }

  RoleDecl parseRole() {
    advance();
    RoleDecl r;
    r.name = expect(TK::Ident).value;
    r.team = r.name;
    expect(TK::Symbol, "{");
    while (!check(TK::Symbol, "}")) {
      std::string field = advance().value;
      expect(TK::Symbol, ":");
      if (field == "team") {
        r.team = expect(TK::String).value;
      } else if (field == "memory") {
        r.memoryName = expect(TK::Ident).value;
        if (match(TK::Symbol, "(")) {
          while (!check(TK::Symbol, ")")) {
            r.memoryArgs.push_back(parseExpr());
            if (!match(TK::Symbol, ",")) break;
          }
          expect(TK::Symbol, ")");
        }
      } else if (field == "sees") {
        r.sees = advance().value;
      } else if (field == "count") {
        if (peek().kind == TK::Ident && peek().value == "remainder") {
          advance();
          r.count = -1;
        } else {
          r.count = intLit();
        }
      } else {
        fail("unknown role field '" + field + "'");
      }
      match(TK::Symbol, ",");
    }
    expect(TK::Symbol, "}");
    return r;
  }

  std::unique_ptr<WorldDecl> parseWorld() {
    advance();
    expect(TK::Symbol, "{");
    auto w = std::make_unique<WorldDecl>();
    while (!check(TK::Symbol, "}")) {
      if (peek().kind == TK::Ident && peek().value == "location") {
        advance();
        LocationTypeDecl lt;
        lt.name = expect(TK::Ident).value;
        expect(TK::Symbol, "{");
        while (!check(TK::Symbol, "}")) {
          std::string field = advance().value;
          expect(TK::Symbol, ":");
          if (field == "tag") {
            lt.tag = expect(TK::String).value;
            lt.hasTag = true;
          } else if (field == "capacity") {
            lt.capacity = intLit();
          } else if (field == "count") {
            lt.count = intLit();
          } else {
            fail("unknown location field '" + field + "'");
          }
          match(TK::Symbol, ",");
        }
        expect(TK::Symbol, "}");
        w->types.push_back(std::move(lt));
        continue;
      }
      std::string field = advance().value;
      expect(TK::Symbol, ":");
      if (field == "width") w->width = intLit();
      else if (field == "height") w->height = intLit();
      else fail("unknown world field '" + field + "'");
      match(TK::Symbol, ",");
    }
    expect(TK::Symbol, "}");
    return w;
  }

  std::unique_ptr<Block> parseBlock() {
    expect(TK::Symbol, "{");
    auto b = std::make_unique<Block>();
    while (!check(TK::Symbol, "}")) {
      if (peek().kind == TK::End) fail("unexpected end of file (missing '}')");
      b->stmts.push_back(parseStatement());
    }
    expect(TK::Symbol, "}");
    return b;
  }

  StmtP stmt(SK k, int line) {
    auto s = std::make_unique<Stmt>();
    s->k = k;
    s->line = line;
    return s;
  }

  StmtP parseStatement() {
    int line = peek().line;
    if (isKw("let")) {
      advance();
      auto s = stmt(SK::Let, line);
      s->sym = prog_.syms.intern(expect(TK::Ident).value);
      expect(TK::Symbol, "=");
      s->e = parseExpr();
      return s;
    }
    if (isKw("if")) return parseIf();
    if (isKw("while")) {
      advance();
      auto s = stmt(SK::While, line);
      s->e = parseExpr();
      s->body = parseBlock();
      return s;
    }
    if (isKw("for")) {
      advance();
      auto s = stmt(SK::For, line);
      s->sym = prog_.syms.intern(expect(TK::Ident).value);
      expect(TK::Keyword, "in");
      s->e = parseExpr();
      s->body = parseBlock();
      return s;
    }
    if (isKw("return")) {
      advance();
      auto s = stmt(SK::Return, line);
      if (!check(TK::Symbol, "}")) s->e = parseExpr();
      return s;
    }
    if (isKw("break")) {
      advance();
      return stmt(SK::Break, line);
    }
    if (isKw("run")) {
      advance();
      auto s = stmt(SK::Run, line);
      s->phaseName = expect(TK::Ident).value;
      return s;
    }
    auto expr = parseExpr();
    if (match(TK::Symbol, "=")) {
      auto s = stmt(SK::Assign, line);
      s->target = std::move(expr);
      s->e = parseExpr();
      return s;
    }
    auto s = stmt(SK::Expr, line);
    s->e = std::move(expr);
    return s;
  }

  StmtP parseIf() {
    int line = peek().line;
    advance();
    auto s = stmt(SK::If, line);
    s->e = parseExpr();
    s->body = parseBlock();
    if (match(TK::Keyword, "else")) {
      if (isKw("if")) s->elseIf = parseIf();
      else s->elseBlock = parseBlock();
    }
    return s;
  }

  ExprP binary(Op op, ExprP l, ExprP r, int line) {
    auto e = node(EK::Binary, line);
    e->op = op;
    e->a = std::move(l);
    e->b = std::move(r);
    return e;
  }

  ExprP parseExpr() { return parseOr(); }

  ExprP parseOr() {
    auto l = parseAnd();
    while (check(TK::Keyword, "or")) {
      int line = advance().line;
      l = binary(Op::Or, std::move(l), parseAnd(), line);
    }
    return l;
  }

  ExprP parseAnd() {
    auto l = parseNot();
    while (check(TK::Keyword, "and")) {
      int line = advance().line;
      l = binary(Op::And, std::move(l), parseNot(), line);
    }
    return l;
  }

  ExprP parseNot() {
    if (check(TK::Keyword, "not")) {
      auto e = node(EK::Unary, advance().line);
      e->op = Op::Not;
      e->a = parseNot();
      return e;
    }
    return parseComparison();
  }

  ExprP parseComparison() {
    static const std::unordered_map<std::string, Op> kOps = {
        {"==", Op::Eq}, {"!=", Op::Ne}, {"<", Op::Lt}, {">", Op::Gt}, {"<=", Op::Le}, {">=", Op::Ge}};
    auto l = parseAdditive();
    while (peek().kind == TK::Symbol) {
      auto it = kOps.find(peek().value);
      if (it == kOps.end()) break;
      int line = advance().line;
      l = binary(it->second, std::move(l), parseAdditive(), line);
    }
    return l;
  }

  ExprP parseAdditive() {
    auto l = parseMultiplicative();
    while (check(TK::Symbol, "+") || check(TK::Symbol, "-")) {
      const Token& t = advance();
      Op op = t.value == "+" ? Op::Add : Op::Sub;
      l = binary(op, std::move(l), parseMultiplicative(), t.line);
    }
    return l;
  }

  ExprP parseMultiplicative() {
    auto l = parseUnary();
    while (check(TK::Symbol, "*") || check(TK::Symbol, "/") || check(TK::Symbol, "%")) {
      const Token& t = advance();
      Op op = t.value == "*" ? Op::Mul : t.value == "/" ? Op::Div : Op::Mod;
      l = binary(op, std::move(l), parseUnary(), t.line);
    }
    return l;
  }

  ExprP parseUnary() {
    if (check(TK::Symbol, "-")) {
      auto e = node(EK::Unary, advance().line);
      e->op = Op::Neg;
      e->a = parseUnary();
      return e;
    }
    return parsePostfix();
  }

  ExprP parsePostfix() {
    auto expr = parsePrimary();
    for (;;) {
      int line = peek().line;
      if (match(TK::Symbol, ".")) {
        const Token& t = peek();
        if (t.kind != TK::Ident && t.kind != TK::Keyword) fail("expected an attribute name after '.'");
        auto e = node(EK::Attr, line);
        e->a = std::move(expr);
        e->sym = prog_.syms.intern(advance().value);
        expr = std::move(e);
      } else if (match(TK::Symbol, "(")) {
        if (expr->k != EK::Ident) fail("only named functions can be called");
        auto e = node(EK::Call, line);
        e->sym = expr->sym;
        while (!check(TK::Symbol, ")")) {
          if (peek().kind == TK::Ident && peek(1).kind == TK::Symbol && peek(1).value == "=") {
            int key = prog_.syms.intern(advance().value);
            advance();
            e->kwargs.emplace_back(key, parseExpr());
          } else {
            e->items.push_back(parseExpr());
          }
          if (!match(TK::Symbol, ",")) break;
        }
        expect(TK::Symbol, ")");
        expr = std::move(e);
      } else if (match(TK::Symbol, "[")) {
        auto e = node(EK::Index, line);
        e->a = std::move(expr);
        e->b = parseExpr();
        expect(TK::Symbol, "]");
        expr = std::move(e);
      } else {
        break;
      }
    }
    return expr;
  }

  ExprP parsePrimary() {
    const Token& t = peek();
    int line = t.line;
    if (t.kind == TK::Number) {
      auto e = node(EK::Num, line);
      e->constant = Value::num(std::stod(advance().value));
      return e;
    }
    if (t.kind == TK::String) {
      auto e = node(EK::Str, line);
      e->constant = Value::str(advance().value);
      return e;
    }
    if (t.kind == TK::Keyword && (t.value == "true" || t.value == "false")) {
      auto e = node(EK::Bool, line);
      e->constant = Value::boolean(advance().value == "true");
      return e;
    }
    if (t.kind == TK::Keyword && t.value == "null") {
      advance();
      return node(EK::Null, line);
    }
    if (t.kind == TK::Ident) {
      auto e = node(EK::Ident, line);
      e->sym = prog_.syms.intern(advance().value);
      return e;
    }
    if (match(TK::Symbol, "(")) {
      auto inner = parseExpr();
      expect(TK::Symbol, ")");
      return inner;
    }
    if (match(TK::Symbol, "[")) {
      auto e = node(EK::List, line);
      while (!check(TK::Symbol, "]")) {
        e->items.push_back(parseExpr());
        if (!match(TK::Symbol, ",")) break;
      }
      expect(TK::Symbol, "]");
      return e;
    }
    if (match(TK::Symbol, "{")) {
      auto e = node(EK::Dict, line);
      while (!check(TK::Symbol, "}")) {
        auto key = parseExpr();
        expect(TK::Symbol, ":");
        auto value = parseExpr();
        e->pairs.emplace_back(std::move(key), std::move(value));
        if (!match(TK::Symbol, ",")) break;
      }
      expect(TK::Symbol, "}");
      return e;
    }
    fail("unexpected token '" + t.value + "'");
  }
};

// ---- Link pass: resolve call sites and `run` statements once.

struct Linker {
  Program& prog;
  std::unordered_map<std::string, int> builtins, fns, phases;

  void expr(Expr* e) {
    if (!e) return;
    if (e->k == EK::Call) {
      const std::string& name = prog.syms.name(e->sym);
      if (auto it = builtins.find(name); it != builtins.end()) e->builtin = it->second;
      else if (auto jt = fns.find(name); jt != fns.end()) e->userFn = jt->second;
      // Neither: left unlinked, reported as "undefined function" only if the
      // call actually executes -- same as the reference interpreters.
    }
    expr(e->a.get());
    expr(e->b.get());
    for (auto& i : e->items) expr(i.get());
    for (auto& [k, v] : e->pairs) { expr(k.get()); expr(v.get()); }
    for (auto& [k, v] : e->kwargs) expr(v.get());
  }
  void stmt(Stmt* s) {
    if (!s) return;
    if (s->k == SK::Run) {
      auto it = phases.find(s->phaseName);
      s->phase = it == phases.end() ? -1 : it->second;
    }
    expr(s->e.get());
    expr(s->target.get());
    block(s->body.get());
    stmt(s->elseIf.get());
    block(s->elseBlock.get());
  }
  void block(Block* b) {
    if (!b) return;
    for (auto& s : b->stmts) stmt(s.get());
  }
  void run() {
    for (int i = 0; i < B_COUNT_; ++i) builtins.emplace(std::string(kBuiltinNames[i]), i);
    for (size_t i = 0; i < prog.fns.size(); ++i) fns.emplace(prog.fns[i].name, static_cast<int>(i));
    for (size_t i = 0; i < prog.phases.size(); ++i) phases.emplace(prog.phases[i].name, static_cast<int>(i));
    for (auto& f : prog.fns) block(f.body.get());
    for (auto& m : prog.memoryDecls) block(m.body.get());
    for (auto& p : prog.phases) block(p.body.get());
    for (auto& r : prog.roles)
      for (auto& a : r.memoryArgs) expr(a.get());
    block(prog.winCondition.get());
    block(prog.loop.get());
  }
};

}  // namespace

std::shared_ptr<Program> parseProgram(const std::string& source) {
  auto prog = std::make_shared<Program>();
  Parser(tokenize(source), *prog).parseProgram();
  Linker{*prog}.run();
  return prog;
}

}  // namespace sl
